ext/drx/drx.c

```diff
@@ -2439,22 +2454,45 @@ drx_expand_scatter_gather(void *drcontext, instrlist_t *bb, OUT bool *expanded)
             scratch_xmm != reg_resize_to_opsz(sg_info.gather_dst_reg, OPSZ_16))
             break;
     }
-    /* Spill the scratch xmm.
-     * TODO i#3844: drreg does not support spilling xmm regs yet, so we do it ourselves.
+    /* Spill the scratch mm reg. We spill the largest reg corresponding to scratch_xmm
+     * that is supported by the system. This is required because mov-ing a part of a
+     * ymm/zmm reg zeroes the remaining automatically. So we need to save the complete
+     * ymm/zmm reg and not just the lower xmm bits.
+     * TODO i#3844: drreg does not support spilling mm regs yet, so we do it ourselves.
      * When that support is available, replace the following with the required drreg API
      * calls.
      */
-    per_thread_t *pt = get_tls_data(drcontext);
-    instrlist_insert_mov_immed_ptrsz(
-        drcontext, ALIGN_FORWARD(pt->scratch_xmm_spill_slot, XMM_ALIGNMENT),
-        opnd_create_reg(scratch_reg0), bb, sg_instr, NULL, NULL);
-    uint mov_xmm_opcode = mov_xmm_aligned32_opcode();
+    int mov_scratch_mm_opcode;
+    opnd_size_t mov_scratch_mm_opnd_sz;
+    reg_id_t scratch_mm;
+    get_mov_scratch_mm_opcode_and_size(&mov_scratch_mm_opcode, &mov_scratch_mm_opnd_sz);
+    scratch_mm = reg_resize_to_opsz(scratch_xmm, mov_scratch_mm_opnd_sz);
+
+    drmgr_insert_read_tls_field(drcontext, tls_idx, bb, sg_instr, scratch_reg0);
     instrlist_meta_preinsert(
         bb, sg_instr,
-        instr_create_1dst_1src(
-            drcontext, mov_xmm_opcode,
-            opnd_create_base_disp(scratch_reg0, DR_REG_NULL, 0, 0, OPSZ_16),
-            opnd_create_reg(scratch_xmm)));
+        INSTR_CREATE_mov_ld(
+            drcontext, opnd_create_reg(scratch_reg0),
+            OPND_CREATE_MEMPTR(scratch_reg0,
+                               offsetof(per_thread_t, scratch_mm_spill_slot))));
+
+    if (mov_scratch_mm_opnd_sz == OPSZ_64) {
+        instrlist_meta_preinsert(
+            bb, sg_instr,
+            instr_create_1dst_2src(drcontext, mov_scratch_mm_opcode,
+                                   opnd_create_base_disp(scratch_reg0, DR_REG_NULL, 0, 0,
+                                                         mov_scratch_mm_opnd_sz),
+                                   opnd_create_reg(DR_REG_K0),
```

Comment:
Is K0 holding a known value here?

Reply:
K0 is a special mask reg. It denotes unmasked operations when used as a mask reg.

Reply:
Added a comment.